src/js/oncoprintmodel.js

```diff
@@ -1450,54 +1451,109 @@ var OncoprintModel = (function () {
 
 var PrecomputedComparator = (function() {
     function PrecomputedComparator(list, comparator, sort_direction, element_identifier_key) {
-	var preferred, mandatory;
-	if (typeof comparator === "function") {
-	    preferred = comparator;
-	    mandatory = comparator;
-	} else {
-	    preferred = comparator.preferred;
-	    mandatory = comparator.mandatory;
-	}
-	var makeDirectedComparator = function(cmp) {
-	    return function (d1, d2) {
-		if (sort_direction === 0) {
-		    return 0;
-		}
-		var res = cmp(d1, d2);
-		if (res === 2) {
-		    return 1;
-		} else if (res === -2) {
-		    return -1;
+		if (typeof comparator === "object" && comparator.vector_length) {
+			initializeVector(this, list, comparator, sort_direction, element_identifier_key);
 		} else {
-		    return res * sort_direction;
+			initializeComparator(this, list, comparator, sort_direction, element_identifier_key);
 		}
-	    };
-	};
-	var preferredComparator = makeDirectedComparator(preferred);
-	var mandatoryComparator = makeDirectedComparator(mandatory);
-	var sorted_list = list.sort(preferredComparator);
-	
-	// i is a change point iff comp(elt[i], elt[i+1]) !== 0
-	this.preferred_change_points = []; // i is a preferred change pt iff its a change pt with comp = preferredComparator but not with comp = mandatoryComparator
-	this.mandatory_change_points = []; // i is a mandatory change pt iff its a change pt with comp = mandatoryComparator
-	
-	// note that by the following process, preferred_change_points and mandatory_change_points are sorted
-	for (var i=0; i<sorted_list.length; i++) {
-	    if (i === sorted_list.length - 1) {
-		break;
-	    }
-	    if (mandatoryComparator(sorted_list[i], sorted_list[i+1]) !== 0) {
-		this.mandatory_change_points.push(i);
-	    } else if (preferredComparator(sorted_list[i], sorted_list[i+1]) !== 0) {
-		this.preferred_change_points.push(i);
-	    }
-	}
-	// Note that by this process change_points is sorted
-	this.id_to_index = {};
-	for (var i=0; i<sorted_list.length; i++) {
-	    this.id_to_index[sorted_list[i][element_identifier_key]] = i;
+    }
+
+    function initializeComparator(precomputed_comparator, list, comparator, sort_direction, element_identifier_key) {
```

Comment:
add comment explaining what this is, what initializeVector is, and what the arguments are